File: eval_fsf.py
import math
import matplotlib.pyplot as plt
import numpy as np
from scipy import signal

class eval_FSF:
    """
    Purpose
    -------
    eval_FSF is a frequency slope fit distortion detector that uses a linear fit of the power in certain frequency bands to measure distortion.
        eval_FSF()
            Creates a frequency slope fit detector using AI bands.
        eval_FSF('Mel')
            Same as above but uses bands from the Mel scale.    
        eval_FSF(band_matrix)
            Same as above but usees bands given in the two column matrix, band_matrix.
    
    Attributes
    ----------
    name : str
        Method name.
    Freq_Set : numpy array
        Frequency bands to use.
    FFT_len : int
        Length of FFT for band power measurments.
    bandX : bool
        Uses band numbers for x values in linear fit.
    
    Methods
    -------
    process(rec, y, clipi) 
        Processes recordings to obtain a distortion estimation.
    slopeCalc(dat)
        Calculates the slope of a given recording.
    filterPlot()
        Plots filter band edges.
    
    """
    
    FFT_len = 2 ** 14
    bandX = True

    def __init__(self, bands='AI'):
        """
        Purpose
        -------
        Constructor for eval_FSF class
        
        Parameters
        ----------
        band : str or numpy array
        
        """
       
        #Check for AI bands
        if bands == 'AI':
            self.Freq_Set = self.AI_bands()
        #Check for Mel bands
        elif bands == 'Mel':
            self.Freq_Set = self.Mel_bands()
        #Check for custom ((numeric bands))
        elif isinstance(bands, np.ndarray): #find a way to check the dtype
            self.Freq_Set = bands
        #Else throw error
        else:
            raise ValueError("Expected either strings 'AI' or 'Mel' or two column numpy array of type double")
            
    def process(self, rec, y, clipi):
        """
        Purpose
        -------
        Calculates a harmonic distortion score from a set of recordings.

        Parameters
        ----------
        rec : list of numpy arrays
            Set of recordings.
        y : list of numpy arrays
            Playback waveforms.
        clipi : numpy array
            Playback waveform index for each recording.

        Returns
        -------
        score : numpy array
            Harmonic distortion score.

        """
        y_slope = np.zeros(sorted(y.shape))
        y_intercept = np.zeros(sorted(y.shape))
        
        for k in range(y.shape[1]):
            y_slope[k], y_intercept[k] = self.slope_calc(y[k])
        
        slope = np.zeros(sorted(clipi.shape))
        intercept = np.zeros(sorted(clipi.shape))
        
        for k in range(rec.shape[1]):
            slope[k], intercept[k] = self.slope_calc(rec[k])
            
        score = np.true_divide(slope, y_slope[clipi])
        
        return score
         
    def slope_calc(self, dat):
        """
        Purpose
        -------
        Calculates the frequency slope fit for a given recording.
        Performs a linear fit of power in a given band.

        Parameters
        ----------
        dat : numpy array
            Playback waveform or recording.

        Returns
        -------
        slope : double
            Slope from linear fit.
        intercept : double
            Intercept from linear fit.

        """
        num_bands = self.Freq_Set.shape[0]
        
        dat = dat.reshape(dat.size, 1)
        
        win = signal.windows.hamming(dat.shape[1])
        freq, pxx = signal.periodogram(dat, fs=48E3, window=win, nfft=self.FFT_len, scaling='spectrum')
        
        band_vals = np.zeros((num_bands, 1))
        
        for band in range(num_bands):
            mask = np.logical_and((freq >= self.Freq_Set[band, 0]), (freq <= self.Freq_Set[band, 1]))
            band_vals[band] = 10 * math.log10(np.mean(pxx[mask]))
            
        max_idx = np.argmax(band_vals[0:round(num_bands/2)])
        
        fit_range = np.arange(max_idx, band_vals.shape[0])
        
        if self.bandX:
            x_vals = np.arange(1, num_bands)
            x_vals = x_vals[fit_range]
            
        else:
            band_center = np.mean(self.Freq_Set, axis=1)
            x_vals = 10 * math.log10(band_center[fit_range])
            
        p = np.polynomial.polynomial.polyfit(x_vals, band_vals[fit_range], 1)
       
        slope =  p[1]
        intercept = p[0]
        
        return slope, intercept
    
    #def filter_plot(self):
     #   num_bands = self.Freq_Set.shape[0]
      #  fig, ax = plt.subplots()
        
    def AI_bands(): 
        """
        Purpose
        -------
        Generates AI bands.

        Returns
        -------
        bands : numpy array
            AI bands.

        """
        bands = np.array([[200, 450],
                          [400, 650],
                          [600, 850],
                          [800, 1050],
                          [1000, 1250],
                          [1200, 1450],
                          [1400, 1650],
                          [1600, 1850],
                          [1800, 2050],
                          [2000, 2250],
                          [2200, 2450],
                          [2400, 2650],
                          [2600, 2850],
                          [2800, 3050],
                          [3000, 3250]])
        return bands
    
    def Mel_bands():
        """
        Purpose
        -------
        Generates Mel bands.

        Returns
        -------
        bands : numpy array
            Mel bands.

        """
        bands = np.array([[267, 400],
                          [333, 467], 
                          [400, 533],
                          [467, 600],
                          [533, 667],
                          [600, 733],
                          [667, 800],
                          [733, 867],
                          [800, 933],
                          [867, 999],
                          [933, 1071],
                          [999, 1147],
                          [1071, 1229],
                          [1147, 1316],
                          [1229, 1410],
                          [1316, 1510],
                          [1410, 1618],
                          [1510, 1733],
                          [1618, 1856],
                          [1733, 1988],
                          [1856, 2130],
                          [1988, 2281], 
                          [2130, 2444],
                          [2281, 2618],
                          [2444, 2804],
                          [2618, 3004],
                          [2804, 3217]])
        return bands

File: eval_FSF Test Cases/eval_FSF.m
classdef eval_FSF < method_eval
%EVAL_FSF - frequency slope fit distortion detector 
%
%   m = EVAL_FSF() creates a frequency slope fit detector using Uniform bands
%
%   The EVAL_FSF class is a method_eval class that uses a linear fit of the
%   power in certain frequency bands to measure distortion.
%
%EVAL_FSF Methods:
%   process    - process recordings to get a distortion estimation
%   filterPlot - plot filter band edges
%   slopeCalc  - calculate slope of a given recording
%
%EVAL_FSF Properties:
%   name      -  method name. 
%   Freq_Set  -  Frequency bands to use
%   FFT_len   -  Length of FFT for band power measurements
%   bandX     -  Use band numbers for x values in fit
%
%   See also EVALTEST, METHOD_EVAL.
%
    properties
%FREQ_SET - The set of frequency band edges as a two column matrix. The
%           first column represents the lower band edges and the second
%           column represents the upper band edges.
        Freq_Set(:,2)=[200 450; 400 650; 600 850; 800 1050; 1000 1250;
                       1200 1450; 1400 1650; 1600 1850; 1800 2050;
                       2000 2250; 2200 2450; 2400 2650; 2600 2850;
                       2800 3050; 3000 3250;];
%FFT_LEN - the FFT length to use for the band power calculations               
       FFT_len=2^14;
%BANDX - logical value that determines if the linear fit x values come from
%        band centers (false) or if band numbers are used (true)
        bandX(1,1) {mustBeNumericOrLogical}=true;
    end
    methods
        function score=process(obj,rec,y,clipi,~)
%PROCESS - calculate a harmonic distortion score from a set of recordings
%
%   score=PROCESS(rec,y,clipi) - return a harmonic distortion score for a
%                                given set of recordings (rec), original
%                                clips (y), and clip indicies (clipi).

            y_slope=zeros(sort(size(y)));
            y_int=zeros(sort(size(y)));
            
            for k=1:length(y)
                [y_slope(k),y_int(k)]=obj.slopeCalc(y{k});
            end

            %preallocate! for speed
            slope=zeros(sort(size(clipi)));
            intercept=zeros(sort(size(clipi)));
            for k=1:length(rec)
                [slope(k),intercept(k)]=obj.slopeCalc(rec{k});
            end
            %compute score for each trial
            score=slope./y_slope(clipi);

        end
        function [slope,intercept]=slopeCalc(obj,dat)
%SLOPECALC - calculate the frequency slope fit for a given recording
%
%   this is an internal method that does a linear fit of the power in a
%   given band and returns the slope, intercept and linear model

            %number of bands
            nbands=size(obj.Freq_Set,1);
            
            %force column vector for dat
            dat=reshape(dat,[],1);

            %calculate periodogram
            [pxx,freq]=periodogram(dat,hamming(length(dat)),obj.FFT_len,48e3,'power');

            %preallocate
            band_val=zeros(nbands,1);

            %calculate data for frequency bands
            for band=1:nbands
                band_val(band)=10*log10(mean(pxx( (freq>=obj.Freq_Set(band,1)) & (freq<=obj.Freq_Set(band,2)) )));
            end

            %get the index of maximum value
            %only consider the first half of the FI bands
            [~,max_i]=max(band_val(1:round(nbands/2)));

            %range of fit is to the right of max
            rng=max_i:length(band_val);

            if(obj.bandX)
                %x-axis values from band number
                xvals=(1:size(obj.Freq_Set,1))';
                %eliminate unused bands
                xvals=xvals(rng);
            else
                %center of each band
                bcenter=mean(obj.Freq_Set,2); 
                %x-axis values from log10 of band centers
                xvals=10*log10(bcenter(rng));
            end
            
            %get linear fit of data
            p=polyfit(xvals,band_val(rng),1);

            %find slope for data
            slope=p(1);
            intercept=p(2);
        end
        
        function filterPlot(obj)
%FILTERPLOT - plot the filter band edges
%
%   FILTERPLOT() makes a plot of the filter bands. The filter bands are
%                represented by rectangles with a x-axis span given by the
%                band edges and the y-axis location corresponding to the
%                band number.

            %number of bands
            nbands=size(obj.Freq_Set,1);
            
            figure;
            hold on;
            filt_colors=lines(nbands);
            %calculate data for frequency bands
            for band=1:nbands
                %create a bar for each filter
                fill(obj.Freq_Set(band,[1 1 2 2]),[1,1,1,1]*band+[-0.4,0.4,0.4,-0.4],filt_colors(band,:),'EdgeColor','none','DisplayName',sprintf('Band #%u',band))
            end
            hold off;
            
            xlabel('Frequency [Hz]');
            ylabel('Filter #');
            
        end
        
    end
    
end